### PR TITLE
Show review labels in the pull request lists

### PR DESCRIPTION
**Pull request by oscarcosta**
Show review labels (Review required, Approved and Changes requested) in the pull request lists. Related to issue #8257 

Screenshot from 2019-12-06 17-56-29

Screenshot from 2019-12-06 17-56-41

Screenshot from 2019-12-06 17-56-52

### CODE REVIEW

**Review by bagasme — COMMENTED** [review]
Related strings for some locales are same as on `en-US` locale.

**Comment by lunny** [comment]
@oscarcosta If you can show approved count, that's better! I found github didn't show that.

File: models/pull.go
Comment (lafriks) [inline models/pull.go]: This check should be before if there are enough approvals as there could be enough approvals but if there is any rejected review it should still be as rejected

File: options/locale/locale_cs-CZ.ini
Comment (lafriks) [inline options/locale/locale_cs-CZ.ini]: Do not add translations to other languages, if there is no translation it will fallback to English anyway

**Review by lafriks — CHANGES_REQUESTED** [review]

File: options/locale/locale_cs-CZ.ini
Comment (oscarcosta) [reply options/locale/locale_cs-CZ.ini]: Got it, thanks for the explanation.

**Comment by oscarcosta** [comment]
Review labels with counters:
Screenshot from 2019-12-07 00-06-27

File: models/pull.go
Comment (guillep2k) [inline models/pull.go]: I think that a PR that has been repeatedly approved and rejected from one reviewer should count only one approval or one rejection, depending on which happened last:

* **Reviewer 1**: approves, requests changes = **0 approvals**, **1 change requests**
* **Reviewer 2**: approves = **1 approval**, **0 change requests**
* **Reviewer 3**: requests changes, requests changes, requests changes, approves = **1 approvals**, **0 change requests**
* **Reviewer 4**: requests changes = **0 approvals**, **1 change requests**

**Grand total: 2 approvals, 2 change requests**
 
To achieve that, IMHO all counts (approvals, rejections) should be processed in the same database query. Check [this comment](https://github.com/go-gitea/gitea/pull/9055#discussion_r350977864) (and other reviews of mine in that PR) for examples on how to achieve that.

Comment (davidsvantesson) [reply models/pull.go]: I think that is already achieved. Both `GetGrantedApprovalsCount` and `GetRejectedReviewsCount` only counts reviews that are _official_, and official reviews are only the latest review (approval or rejection) of users that are in the whitelist.

**Review by davidsvantesson — CHANGES_REQUESTED** [review]
In general I think this PR looks good! 
One issue is that we haven't implemented blocking PR on changes requested yet (#9130), so the list will say "Changes requested" although it will not block merging. I think it would be easy to implement that after this PR though.
I think we need to decide how it should work:
 - If changes are requested it will always block merging (no option for it). For me this would be OK and makes sense, though it would be a breaking change. 
 - Have a branch protection option whether requested changes shall block merging. In that case I don't know if it should say "Changes Requested" in the PR list, don't make sense if it isn't blocking?

File: models/pull.go
Comment (guillep2k) [reply models/pull.go]: Duh, I forgot about the `official` property! I had a rough week... I need some rest. 😄

File: routers/user/home.go
Comment (oscarcosta) [reply routers/user/home.go:542]: Thanks for the feedback. ```Issue.LoadPullRequest()``` was being called only to set the reference of Issue in the PullRequest. I changed it to set the Issue reference when the PullRequests are loaded.

File: options/locale/locale_es-ES.ini
Comment (oscarcosta) [reply options/locale/locale_es-ES.ini]: OK, changed id. Thanks.

File: templates/user/dashboard/issues.tmpl
Comment (guillep2k) [inline templates/user/dashboard/issues.tmpl]:
```suggestion
									<a class="milestone" href="{{AppSubUrl}}/{{.Repo.Owner.Name}}/{{.Repo.Name}}/pulls/{{.Index}}">{{.PullRequest.GetReviewLabel $.i18n.Lang | Safe}}</a>
```

File: models/issue.go
Comment (guillep2k) [inline models/issue.go]: Any reason to not simplify it like this?

```suggestion
	if issue.PullRequest != nil {
```

File: models/pull.go
Comment (lunny) [reply models/pull.go]: We could use different color tag for approved counter and requested counter. Even the approval requirement. `1/2` means this PR require 2 approvals but has 1 approved.

Comment (lunny) [reply models/pull.go]: One should approve or request change.

File: templates/repo/issue/list.tmpl
Comment (oscarcosta) [reply templates/repo/issue/list.tmpl:242]: I've tried that but got an error. It seems it works in this way:
```{{if and .IsPull .PullRequest.RequiresApproval}}```
I'm not sure what the parentheses change in the expression.

File: models/issue.go
Comment (oscarcosta) [reply models/issue.go]: I just didn't want to change the reference if it's already set.

File: models/pull.go
Comment (oscarcosta) [reply models/pull.go]: I've tried to display these counters after the label. Does it seems right?
Screenshot from 2019-12-08 11-22-35

File: templates/repo/issue/list.tmpl
Comment (guillep2k) [reply templates/repo/issue/list.tmpl:242]: As long as it works. I've used `or(...)` before, so I thought it was the same with `and`. Go figure. 😄

Comment (oscarcosta) [reply templates/repo/issue/list.tmpl:242]: Actually, ```{{if and .IsPull .PullRequest.RequiresApproval}}``` broke the build. It seems ```and exp1 exp2``` tries to evaluate both expressions. Sorry, Go and its templates are pretty new for me.

File: models/pull.go
Comment (guillep2k) [reply models/pull.go]: It's incomplete information. The PR can have 17 approvals and 1 change request. Should it show "1 changes requested" or "17 people approved this"? I think both.

File: templates/repo/issue/list.tmpl
Comment (guillep2k) [reply templates/repo/issue/list.tmpl:242]: Oh, yes, `if` in go templates have no shortcut expressions. Sorry, I didn't remember that when I made the suggestion. 😳

File: models/issue.go
Comment (6543) [reply models/issue.go]: @oscarcosta why not use `.LoadIssue()` it does exact this

File: models/issue_list.go
Comment (6543) [inline models/issue_list.go]:
```suggestion
			issue.PullRequest.LoadIssue()
```

File: models/issue.go
Comment (6543) [reply models/issue.go]: I would not add this in here! (issue.PR.issue = issue) sounds like a bad idear!
just pull.LoadIssue() when you need it!

File: models/branches.go
Comment (6543) [inline models/branches.go]: can you add pr.LoadIssue() before?

File: models/issue_list.go
Comment (6543) [inline models/issue_list.go]: same as here https://github.com/go-gitea/gitea/pull/9274/files#r373856895

**Comment by 6543** [comment]
@oscarcosta can you resolve change-requests and conflicts pleace

**Comment by oscarcosta** [comment]
@6543 sure, I can work on that today.

File: models/branches.go
Comment (oscarcosta) [reply models/branches.go]: I wondering if would not be better use `pr.IssueID` instead of `pr.Issue.ID`, likewise the other methods.

Comment (jolheiser) [reply models/branches.go]: Yes, `pr.IssueID` is the way to go.

**Review by oscarcosta — COMMENTED** [review]
Reviewed the changes proposed by @6543

File: models/branches.go
Comment (lunny) [inline models/branches.go]: The method could be a method or `PullRequest`.

File: models/issue.go
Comment (lunny) [inline models/issue.go]: Why we need reload issue?

File: models/issue.go
Comment (6543) [reply models/issue.go]: this was not exactly what I ment ... I'll send a pull-request ...

**Review by lunny — CHANGES_REQUESTED** [review]

**Comment by 6543** [comment]
@oscarcosta just merge https://github.com/oscarcosta/gitea/pull/1

**Review by 6543 — COMMENTED** [review]
for templates lgtm ..  
but for `models/branches.go` and `models/pull.go` I'm not happy ..

File: models/branches.go
Comment (6543) [reply models/branches.go]: @oscarcosta protectBranch is never used so it should not be a function of it ..

File: models/pull.go
Comment (6543) [inline models/pull.go]: translation messages ... shold not be use in models ... so if you need this should look at /router.. ors where you could add the function or services/pull/ ..?
@lunny where do you think should go the stuff witch doesnt touch db ...

Comment (lunny) [reply models/pull.go]: We should give some status on this method and put the language message on tmpl files.

**Review by lafriks — APPROVED** [review]

File: models/pull.go
Comment (lafriks) [inline models/pull.go]: This function should return some kind of status and translation strings should be in template file

**Comment by lunny** [comment]
Two nit things will make this better.
- [ ] like what @lafrks said return a status. And you could also check if `ConflictedFiles` length to return conflicted status.
- [ ] batch load protected branch or other objects on `pull_list.go` for a performance condieration.

**Comment by oscarcosta** [comment]
Hi @lunny,
Wouldn't the conflict status be slightly different from the review status? Perhaps a different label in the pull request list.

**Review by lunny — CHANGES_REQUESTED** [review]
I think this is the last one thought from me. Otherwise LGTM.

File: models/pull.go
Comment (davidsvantesson) [reply models/pull.go:197]: That is in the `official` column which is accounted for, only last review of a reviewer can be official. (The reviewer shall also be in the review whitelist)

Comment (lunny) [reply models/pull.go:197]: ~`official` means this review could be counted as approvals but it don't mean it's the last review of the official reviewer. See https://github.com/go-gitea/gitea/blob/master/models/review.go#L57-L58~

**Comment by lunny** [comment]
We just should use `select status, count(status) from review where ... and is_offcial=true group by status`. Because `is_offcial` means official reviewers' last review.

**Comment by oscarcosta** [comment]
Hi @lunny, 
I tried to load the pull request attributes in batch, calling the methods in the loadAttributes in pull_list.go. However, the PullRequests are loaded as Issues with the isPull attribute setted (see files routers/user/home.go and routers/repo/issue.go), so I called the PullRequest.LoadAttributes in those places.

**Comment by oscarcosta** [comment]
> We just should use `select status, count(status) from review where ... and is_offcial=true group by status`. Because `is_offcial` means official reviewers' last review.

About this query, would it replace both GetRejectedReviewsCount and GetGrantedApprovalsCount methods in GetReviewStatus?

**Comment by 6543** [comment]
Whats the status of this?

**Comment by oscarcosta** [comment]
> Whats the status of this?

Sorry, but I think I will not be able to finish it. If anyone wants to finish it, go ahead.

**Comment by 6543** [comment]
@oscarcosta I couldn't find the specific pull but its arlready implemented:

Bildschirmfoto zu 2020-04-26 15-59-36

**Pull request closed without merging**
